### PR TITLE
[collectd 6] add reference counting to plugins that register multiple callbacks with the same `user_data_t`.

### PR DESCRIPTION
**Pull request by octo**
This outlines an alternative fix to the problem described in #4117.

The idea is to fix the plugins rather than make the plugin's memory management a problem of the shutdown code in `src/daemon/plugin.c`. By adding a reference count, it doesn't matter in which order the various callbacks are shut down in the daemon code.

I hope that the same approach can be used to fix other affected plugins (write_riemann, write_stackdriver, write_syslog, write_tsdb), but I have not verified that yet.

### CODE REVIEW

File: src/write_riemann.c [hunk 1]
```diff
@@ -616,16 +616,16 @@ static void wrr_free(void *p) /* {{{ */
...
 
   pthread_mutex_lock(&host->lock);
-
   host->reference_count--;
-  if (host->reference_count > 0) {
-    pthread_mutex_unlock(&host->lock);
+  int reference_count = host->reference_count;
+  pthread_mutex_unlock(&host->lock);
+
+  if (reference_count > 0) {
```
Comment (eero-t) [inline src/write_riemann.c:623]: What's the reason for this change, doesn't it introduce small chance for race condition?

PS.  See also https://github.com/collectd/collectd/issues/4050

**Review by eero-t — COMMENTED** [review]
On quick look this seems OK, I just have a question about one Rieman plugin change.

I would also suggest running collectd with these plugins under Valgrind memcheck (default valgrind tool) and ones discussed here: https://github.com/collectd/collectd/pull/4026#issuecomment-1214894922.

File: src/write_riemann.c (hunk 1 above)
Comment (octo) [reply src/write_riemann.c]: I wanted to keep this code consistent across plugins. This looks like an oversight. Fixed.

File: src/write_riemann.c [hunk 2]
```diff
@@ -625,7 +624,7 @@ static void wrr_free(void *p) /* {{{ */
 
   wrr_disconnect(host);
 
-  pthread_mutex_lock(&host->lock);
+  pthread_mutex_unlock(&cb->curl_lock);
   pthread_mutex_destroy(&host->lock);
```
Comment (eero-t) [inline src/write_riemann.c:628]: Changing lock to unlock looks like bug fix, but why the `host->lock` to  `cb->curl_lock` change, when only `host->lock` is destroyed (and other plugins destroy `cb` lock)?

**Review by eero-t — APPROVED** [review]
Looks OK to me now, but I would really recommend doing also init/deinit testing under  Valgrind/ASAN to verify at least riemann plugin validity.

**Comment by octo** [comment]
I could not test the *write_riemann plugin* because it hasn't been ported to collectd 6 yet. I did run the *write_http plugin* with Valgrind (both memcheck and helgrind) without any findings.
